exchanges/bitstamp/bitstamp_websocket.go

```diff
@@ -101,137 +117,190 @@ func (b *Bitstamp) wsHandleData(respRaw []byte) error {
+}
 
-		p, err := currency.NewPairFromFormattedPairs(currencyPair, enabledPairs, pFmt)
-		if err != nil {
-			return err
-		}
+func (b *Bitstamp) handleWSOrderbook(_ context.Context, wsResp *websocketResponse, msg []byte) error {
+	if wsResp.pair.IsEmpty() {
+		return errWSPairParsingError
+	}
 
-		err = b.wsUpdateOrderbook(&wsOrderBookTemp.Data, p, asset.Spot)
-		if err != nil {
-			return err
-		}
-	case "trade":
-		if !b.IsSaveTradeDataEnabled() {
-			return nil
-		}
-		wsTradeTemp := websocketTradeResponse{}
-		err := json.Unmarshal(respRaw, &wsTradeTemp)
-		if err != nil {
-			return err
-		}
+	wsOrderBookTemp := websocketOrderBookResponse{}
+	err := json.Unmarshal(msg, &wsOrderBookTemp)
+	if err != nil {
+		return err
+	}
 
-		var currencyPair string
-		splitter := strings.Split(wsResponse.Channel, currency.UnderscoreDelimiter)
-		if len(splitter) == 3 {
-			currencyPair = splitter[2]
-		} else {
-			return errWSPairParsingError
-		}
-		pFmt, err := b.GetPairFormat(asset.Spot, true)
-		if err != nil {
-			return err
-		}
+	return b.wsUpdateOrderbook(&wsOrderBookTemp.Data, wsResp.pair, asset.Spot)
+}
 
-		enabledPairs, err := b.GetEnabledPairs(asset.Spot)
-		if err != nil {
-			return err
-		}
+func (b *Bitstamp) handleWSTrade(_ context.Context, wsResp *websocketResponse, msg []byte) error {
+	if !b.IsSaveTradeDataEnabled() {
+		return nil
+	}
 
-		p, err := currency.NewPairFromFormattedPairs(currencyPair, enabledPairs, pFmt)
-		if err != nil {
-			return err
-		}
+	if wsResp.pair.IsEmpty() {
+		return errWSPairParsingError
+	}
 
-		side := order.Buy
-		if wsTradeTemp.Data.Type == 1 {
-			side = order.Sell
-		}
-		var a asset.Item
-		a, err = b.GetPairAssetType(p)
-		if err != nil {
-			return err
-		}
-		return trade.AddTradesToBuffer(b.Name, trade.Data{
-			Timestamp:    time.Unix(wsTradeTemp.Data.Timestamp, 0),
-			CurrencyPair: p,
-			AssetType:    a,
-			Exchange:     b.Name,
-			Price:        wsTradeTemp.Data.Price,
-			Amount:       wsTradeTemp.Data.Amount,
-			Side:         side,
-			TID:          strconv.FormatInt(wsTradeTemp.Data.ID, 10),
-		})
-	case "order_created", "order_deleted", "order_changed":
-		if b.Verbose {
-			log.Debugf(log.ExchangeSys, "%v - Websocket order acknowledgement", b.Name)
+	wsTradeTemp := websocketTradeResponse{}
+	if err := json.Unmarshal(msg, &wsTradeTemp); err != nil {
+		return err
+	}
+
+	side := order.Buy
+	if wsTradeTemp.Data.Type == 1 {
+		side = order.Sell
+	}
+	return trade.AddTradesToBuffer(b.Name, trade.Data{
+		Timestamp:    time.Unix(wsTradeTemp.Data.Timestamp, 0),
+		CurrencyPair: wsResp.pair,
+		AssetType:    asset.Spot,
+		Exchange:     b.Name,
+		Price:        wsTradeTemp.Data.Price,
+		Amount:       wsTradeTemp.Data.Amount,
+		Side:         side,
+		TID:          strconv.FormatInt(wsTradeTemp.Data.ID, 10),
+	})
+}
+
+func (b *Bitstamp) handleWSOrder(_ context.Context, wsResp *websocketResponse, msg []byte) error {
+	if wsResp.channelType != bitstampAPIWSMyOrders {
+		return nil
```

Comment:
This could return an error if its something that's not set or completely different.

Reply:
I was going to comment to explain WHY we return here.
Then I realised it's an anti-pattern. It's not handleWSOrder's job to know what channel it got an order from, that's delegation of responsibilty.
So I've moved the responsibility up to the caller and added a comment.
See 681dbf21
Resolving assuming you'll be happy but please reopen if not.